Add Perceiver IO

NielsRogge opened the pull request:
# What does this PR do?

This PR implements [Perceiver IO](https://arxiv.org/abs/2107.14795), by Google Deepmind.

Fixes #12996

It's a pretty cool piece of work: it applies a Transformer encoder to any kind of modality (images, text, audio, video) and for any problem (text classification, image classification, audio classification, video autoencoding, optical flow,...)! Perceiver is basically a BERT, ViT and Wav2Vec2 in one. However, the authors did apply the Perceiver on each problem separately, although I believe you could train a single (shared) encoder as backbone, and have different pre- and postprocessors for the different modalities.

The memory- and time requirements of the self-attention mechanism don't depend on the length of the inputs, as the inputs are only used for doing cross-attention. The bulk of computation happens on a set of latent variables, which are just randomly initialized at the beginning of training (i.e. `self.latents = nn.Parameter(torch.randn((num_latents, d_latents)))`).

Some demo notebooks to showcase what Perceiver can do:

- image classification & masked language modeling: [colab](https://colab.research.google.com/drive/1drKjC2EH8YvYAtIayUzmQ82A3KJkf-ka?usp=sharing)
- optical flow: [colab](https://colab.research.google.com/drive/1NE6BKj7JBlNgnntNWoyYOorsgOQCXy-u?usp=sharing)
- video autoencoding (and video classification): [colab](https://colab.research.google.com/drive/1V9BYnF4nK89NNcNOMw3431bw-Q6H8lsP?usp=sharing)

These colab notebooks are based on the original ones provided by Deepmind, in their (very clean) JAX/Haiku implementation which can be found [here](https://github.com/deepmind/deepmind-research/tree/master/perceiver).

cc @esceptico

NielsRogge committed:
First draft

NielsRogge committed:
Style and remove mlm

NielsRogge committed:
Make forward pass work

NielsRogge committed:
More improvements

NielsRogge committed:
More improvements

NielsRogge committed:
Fix bug

NielsRogge committed:
More improvements

NielsRogge committed:
More improvements

NielsRogge committed:
Add PerceiverTokenizer first draft

NielsRogge committed:
Improve conversion script

NielsRogge committed:
More improvements

NielsRogge committed:
Make conversion script work for the encoder

NielsRogge committed:
Make conversion script work with local pickle files

NielsRogge committed:
Style & quality, fix-copies

NielsRogge committed:
Add dummy input to conversion script

NielsRogge committed:
Add absolute position embeddings to TextPreProcessor

NielsRogge committed:
Make forward pass of encoder work

NielsRogge committed:
More improvements

NielsRogge committed:
Move text preprocessor to separate script

NielsRogge committed:
More improvements

NielsRogge committed:
More improvements

NielsRogge committed:
Add post processor

NielsRogge committed:
Make MLM model work

NielsRogge committed:
Style

NielsRogge committed:
Add PerceiverForMaskedLM

NielsRogge committed:
Add PerceiverImagePreprocessor

NielsRogge committed:
Make style

NielsRogge committed:
Make PerceiverForImageClassification work

NielsRogge committed:
More improvements

NielsRogge committed:
More improvements

NielsRogge committed:
Use tokenizer in conversion script

NielsRogge committed:
Use PerceiverForMaskedLM in conversion script

NielsRogge committed:
Define custom PerceiverModelOutput

NielsRogge committed:
Improve PerceiverAttention to make it work for both MLM and image classification

NielsRogge committed:
More improvements

NielsRogge committed:
More improvements

NielsRogge committed:
More improvements to the conversion script

NielsRogge committed:
Make conversion script work for both MLM and image classification

NielsRogge committed:
Add PerceiverFeatureExtractor

NielsRogge committed:
More improvements

NielsRogge committed:
Style and quality

NielsRogge committed:
Add center cropping

NielsRogge committed:
Fix bug

NielsRogge committed:
Small fix

NielsRogge committed:
Add print statement

NielsRogge committed:
Fix bug in image preprocessor

NielsRogge committed:
Fix bug with conversion script

NielsRogge committed:
Make output position embeddings an nn.Parameter layer instead of nn.Embedding

NielsRogge committed:
Comment out print statements

NielsRogge committed:
Add position encoding classes

NielsRogge committed:
More improvements

NielsRogge committed:
Use position_encoding_kwargs

NielsRogge committed:
Add PerceiverForImageClassificationFourier

NielsRogge committed:
Make style & quality

NielsRogge committed:
Add PerceiverForImageClassificationConvProcessing

NielsRogge committed:
Style & quality

NielsRogge committed:
Add flow model

NielsRogge committed:
Move processors to modeling file

NielsRogge committed:
Make position encodings modular

NielsRogge committed:
Make basic decoder use modular position encodings

NielsRogge committed:
Add PerceiverForOpticalFlow to conversion script

NielsRogge committed:
Add AudioPreprocessor

NielsRogge committed:
Make it possible for the basic decoder to use Fourier position embeddings

NielsRogge committed:
Add PerceiverForMultimodalAutoencoding

NielsRogge committed:
Improve model for optical flow

NielsRogge committed:
Improve _build_network_inputs method

NielsRogge committed:
Add print statement

NielsRogge committed:
Fix device issue

NielsRogge committed:
Fix device of Fourier embeddings

NielsRogge committed:
Add print statements for debugging

NielsRogge committed:
Add another print statement

NielsRogge committed:
Add another print statement

NielsRogge committed:
Add another print statement

NielsRogge committed:
Add another print statement

NielsRogge committed:
Improve PerceiverAudioPreprocessor

NielsRogge committed:
Improve conversion script for multimodal modal

NielsRogge committed:
More improvements

NielsRogge committed:
More improvements

NielsRogge committed:
Improve multimodal model

NielsRogge committed:
Make forward pass multimodal model work

NielsRogge committed:
More improvements

NielsRogge committed:
Improve tests

NielsRogge committed:
Fix some more tests

NielsRogge committed:
Add output dataclasses

NielsRogge committed:
Make more tests pass

NielsRogge committed:
Add print statements for debuggin

NielsRogge committed:
Add tests for image classification

NielsRogge committed:
Add PerceiverClassifierOutput

NielsRogge committed:
More improvements

NielsRogge committed:
Make more tests pass for the optical flow model

NielsRogge committed:
Make style & quality

NielsRogge committed:
Small improvements

NielsRogge committed:
Don't support training for optical flow model for now

NielsRogge committed:
Fix _prepare_for_class for tests

NielsRogge committed:
Make more tests pass, add some docs

NielsRogge committed:
Add multimodal model to tests

NielsRogge committed:
Minor fixes

NielsRogge committed:
Fix tests

NielsRogge committed:
Improve conversion script

NielsRogge committed:
Make fixup

NielsRogge committed:
Remove pos_dim argument

NielsRogge committed:
Fix device issue

NielsRogge committed:
Potential fix for OOM

NielsRogge committed:
Revert previous commit

NielsRogge committed:
Fix test_initialization

NielsRogge committed:
Add print statements for debugging

NielsRogge committed:
Fix print statement

NielsRogge committed:
Add print statement

NielsRogge committed:
Add print statement

NielsRogge committed:
Add print statement

NielsRogge committed:
Add print statement

NielsRogge committed:
Add print statement

NielsRogge committed:
Add print statement

NielsRogge committed:
Remove need for output_shape

NielsRogge committed:
Comment out output_shape

NielsRogge committed:
Remove unnecessary code

NielsRogge committed:
Improve docs

NielsRogge committed:
Fix make fixup

NielsRogge committed:
Remove PerceiverTextProcessor from init

NielsRogge committed:
Improve docs

NielsRogge committed:
Small improvement

LysandreJik reviewed (commented):
This looks great, @NielsRogge! This PR is especially clean. I left mostly nits, with a few points I'd like to check with you however:
- The configuration values of the `Preprocessor` objects are hardcoded in the code. I think it would make sense to control these attributes from the configuration object.
- I'm not sure it makes sense to expose these to the end-user. Do they serve a purpose? If so, they should probably have code examples and documentation that explain how they work in detail.
- Have you tried using the image classification models with the image classification example?
- I have personally only played with the Optical flow model, but it seems to have quite a few quirks regarding the input dimensions. Also, there needed to be a lot of methods to compute the optical flow and others, without which the model is quite complex to use. Would it be possible to reference this notebook in the docs (alongside your other notebooks), to add a script in the `research-projects`, or at least to show the full pipeline when using that model? I think that without these, it will be very complex for users to leverage the model which is a shame.

Great work on the perceiver model!

LysandreJik commented on src/transformers/models/perceiver/modeling_perceiver.py:
```diff
@@ -0,0 +1,2726 @@
+# Below: IO pre- and post-processor classes for Perceiver.
+def space_to_depth(frames: torch.Tensor, temporal_block_size: int = 1, spatial_block_size: int = 1) -> torch.Tensor:
+    """Space to depth transform, using einops."""
+    try:
+        import einops
+    except ImportError:
+        raise ImportError("Einops is not installed in your environment, which is required.")
```
I had somehow missed this :eyes: It would really be better to leverage raw `torch` for that

sgugger replied:
Let's not add a new dependency indeed and rely on PyTorch ops.

sgugger commented on docs/source/model_doc/perceiver.rst:
```diff
@@ -0,0 +1,228 @@
+The main problem with the self-attention mechanism of the Transformer is that the time- and memory requirements scale
+quadratically in the sequence length. Hence, models like BERT and RoBERTa are limited to a max sequence length of 512
+tokens. Perceiver aims to solve this issue by, instead of performing self-attention on the inputs, perform it on a set
+of latent variables, and only use the inputs for cross-attention. In this way, the time- and memory requirements don't
```
```suggestion
of latent variables, and only use the inputs for cross-attention. In this way, the time and memory requirements don't
```

sgugger commented on docs/source/model_doc/perceiver.rst:
```diff
@@ -0,0 +1,228 @@
+the :obj:`d_latents` to :obj:`num_labels`.
+
+This was the idea of the original Perceiver paper. However, it could only output classification logits. In their
+follow-up work, PerceiverIO, they generalized it to let the model also produce outputs of arbitrary size. How, you
```
In their follow-up work does not work as they is not defined. Rephrase to "In a follow-up work" for instance.

sgugger commented on docs/source/model_doc/perceiver.rst:
```diff
@@ -0,0 +1,228 @@
+cross-attention with the last hidden states of the latents, using the outputs as queries, and the latents as keys and
+values.
+
+So let's say you want to perform masked language modeling (BERT-style) with the Perceiver. As the Perceiver's input
```
This paragraph mixes you and we. Pick one and stay consistent.

sgugger commented on docs/source/model_doc/perceiver.rst:
```diff
@@ -0,0 +1,228 @@
+    :members: 
+
+
+PerceiverTextPreprocessor
```
Agreed. They can be documented like the specific outputs if you really want to, but they shouldn't be in the main init.

NielsRogge replied:
I believe it can be useful for end-users, because the Perceiver is really mixing and matching lego blocks together. 

For example, one could create a Perceiver for audio classification by defining a new PerceiverModel that uses a PerceiverAudioPreprocessor as input preprocessor and a PerceiverClassificationDecoder as decoder.

The models that are added (PerceiverForImageClassification, PerceiverForOpticalFlow, etc.) just show how you can mix-and-match these pre and postprocessors to make the Perceiver handle various modalities. They aren't as configurable as a ViT or BERT., you cannot easily tweak their attributes without breaking the model. Every Perceiver model is independent and more of an example to showcase what you can do with the Perceiver.

NielsRogge replied:
Ok, that makes sense. And what about documentation, can I include them in the docs?

sgugger commented on src/transformers/__init__.py:
```diff
@@ -599,6 +601,7 @@
     _import_structure["modeling_utils"] = ["Conv1D", "PreTrainedModel", "apply_chunking_to_forward", "prune_layer"]
 
     # PyTorch models structure
+
```
```suggestion
```

sgugger commented on src/transformers/__init__.py:
```diff
@@ -2461,6 +2491,7 @@
         from .utils.dummy_pytorch_quantization_and_torch_objects import *
 
     if is_torch_available():
+
```
```suggestion
```

sgugger commented on src/transformers/models/perceiver/configuration_perceiver.py:
```diff
@@ -0,0 +1,162 @@
+    Configuration objects inherit from :class:`~transformers.PretrainedConfig` and can be used to control the model
+    outputs. Read the documentation from :class:`~transformers.PretrainedConfig` for more information.
+
+
```
```suggestion
```

sgugger commented on src/transformers/models/perceiver/configuration_perceiver.py:
```diff
@@ -0,0 +1,162 @@
+        max_position_embeddings=2048,
+        train_size=[368, 496],
+        num_frames=16,
+        audio_samples_per_frame=1920,  # 48000 // 25
```
I'm not sure if the comment is supposed to be helpful, but given the fact there is no 48000 or 25 in this module, it's rather mysterious -> can be removed.

sgugger commented on src/transformers/models/perceiver/convert_perceiver_haiku_to_pytorch.py:
```diff
@@ -0,0 +1,476 @@
+logger = logging.get_logger(__name__)
+
+# We will verify our results on an image of a dog
+
+
```
```suggestion
```
(Move the comment after the function name if black reintroduces the two blank lines)

sgugger commented on src/transformers/models/perceiver/convert_perceiver_haiku_to_pytorch.py:
```diff
@@ -0,0 +1,476 @@
+
+    state = None
+    if isinstance(checkpoint, dict) and task in [
+        "image_classification",
+        "image_classification_fourier",
+        "image_classification_conv",
```
Fits in one line.

sgugger commented on src/transformers/models/perceiver/convert_perceiver_haiku_to_pytorch.py:
```diff
@@ -0,0 +1,476 @@
+        id2label = {int(k): v for k, v in id2label.items()}
+        config.id2label = id2label
+        config.label2id = {v: k for k, v in id2label.items()}
+        if task == "image_classification":
+            config.image_size = 224
+            model = PerceiverForImageClassification(config)
+        elif task == "image_classification_fourier":
+            config.d_model = 261
+            model = PerceiverForImageClassificationFourier(config)
+        elif task == "image_classification_conv":
+            config.d_model = 322
+            model = PerceiverForImageClassificationConvProcessing(config)
```
I don't think *task* is the appropriate argument to distinguish between models. The task is image classification in all cases, the model should be picked with the architectures field of the config.

NielsRogge replied:
This is just the conversion script, there's no task attribute in the config.

sgugger replied:
Then can we rename this to something different of task?

NielsRogge replied:
Ok, renamed to "architecture".

sgugger commented on src/transformers/models/perceiver/convert_perceiver_haiku_to_pytorch.py:
```diff
@@ -0,0 +1,476 @@
+        raise ValueError(f"Task {task} not supported")
+    model.eval()
+
+    # # multimodal autoencoding requires a dummy input first in order to create all parameters
+    # # (as some parameters are defined in the forward pass)
+    # input_mask = None
+    # if task == "multimodal_autoencoding":
+    #     inputs = dict(image=images, audio=audio, label=torch.zeros((images.shape[0], 700)))
+    #     outputs = model(inputs, attention_mask=input_mask, subsampled_output_points=subsampling)
```
To clean up?

sgugger commented on src/transformers/models/perceiver/feature_extraction_perceiver.py:
```diff
@@ -0,0 +1,180 @@
+    r"""
+    Constructs a Perceiver feature extractor.
+
+    This feature extractor inherits from :class:`~transformers.FeatureExtractionMixin` which contains most of the main
```
```suggestion
    This feature extractor inherits from :class:`~transformers.ImageFeatureExtractionMixin` which contains most of the main
```

sgugger commented on src/transformers/models/perceiver/feature_extraction_perceiver.py:
```diff
@@ -0,0 +1,180 @@
+
+    def __call__(
+        self,
+        images: Union[
+            Image.Image, np.ndarray, "torch.Tensor", List[Image.Image], List[np.ndarray], List["torch.Tensor"]  # noqa
+        ],
```
Define an alias for that type as it makes the signature unreadable.

sgugger commented on src/transformers/models/perceiver/modeling_perceiver.py:
```diff
@@ -0,0 +1,2726 @@
+# limitations under the License.
+""" PyTorch Perceiver model. """
+
+
```
```suggestion
```

sgugger commented on src/transformers/models/perceiver/modeling_perceiver.py:
```diff
@@ -0,0 +1,2726 @@
+from .configuration_perceiver import PerceiverConfig
+
+
+ModalitySizeT = Mapping[str, int]
```
We don't use one-letter variables in the library, so the last T should be expended. If it stands for Type, this should become `ModalitySizeType`.

sgugger commented on src/transformers/models/perceiver/modeling_perceiver.py:
```diff
@@ -0,0 +1,2726 @@
+
+ModalitySizeT = Mapping[str, int]
+PreprocessorOutputT = Tuple[torch.Tensor, Optional[torch.Tensor], torch.Tensor]
+PreprocessorT = Callable[..., PreprocessorOutputT]
+PostprocessorT = Callable[..., Any]
```
Same here

sgugger commented on src/transformers/models/perceiver/modeling_perceiver.py:
```diff
@@ -0,0 +1,2726 @@
+        self.latents = nn.Parameter(torch.randn(config.num_latents, config.d_latents))
+
+    def forward(self, batch_size):
+        embeddings = self.latents.expand(batch_size, -1, -1)  # Thanks, Phil Wang
+
+        return embeddings
```
Can all be done in one line:
```suggestion
        return self.latents.expand(batch_size, -1, -1)  # Thanks, Phil Wang
```

sgugger commented on src/transformers/models/perceiver/modeling_perceiver.py:
```diff
@@ -0,0 +1,2726 @@
+        hidden_states = self.layernorm1(hidden_states)
+        inputs = self.layernorm2(inputs)
+
+        # Project queries, keys and values to a common feature dimension.
+        # If this is instantiated as a cross-attention module, the keys
+        # and values come from the inputs; the attention mask needs to be
+        # such that the inputs's non-relevant tokens are not attended to.
```
We don't have a 60 char limit, so those comments can be expanded horizontally.

sgugger commented on src/transformers/models/perceiver/modeling_perceiver.py:
```diff
@@ -0,0 +1,2726 @@
+        attention_scores = attention_scores / math.sqrt(q_head_dim)
+
+        if attention_mask is not None:
+            # Apply the attention mask is (precomputed for all layers in PerceiverModel forward() function)
```
I don't understand what "Apply the attention mask is" means.

sgugger commented on src/transformers/models/perceiver/modeling_perceiver.py:
```diff
@@ -0,0 +1,2726 @@
+        output_attentions=False,
+    ):
+        self_outputs = self.self(
+            hidden_states,
+            attention_mask,
+            head_mask,
+            inputs,
+            inputs_mask,
+            output_attentions,
```
Fits in one line.

sgugger commented on src/transformers/models/perceiver/modeling_perceiver.py:
```diff
@@ -0,0 +1,2726 @@
+    ):
+        attention_outputs = self.attention(
+            hidden_states,
+            attention_mask,
+            head_mask,
+            inputs,
+            inputs_mask,
+            output_attentions=output_attentions,
```
Either use keyword arguments for all the ones that are optional, or positional arguments for everything.

sgugger commented on src/transformers/models/perceiver/modeling_perceiver.py:
```diff
@@ -0,0 +1,2726 @@
+
+        return outputs
+
+    def feed_forward_chunk(self, attention_output):
+        layer_output = self.layernorm(attention_output)
+
+        layer_output = self.mlp(layer_output)
+
+        return layer_output
```
```suggestion
    def feed_forward_chunk(self, attention_output):
        layer_output = self.layernorm(attention_output)
        layer_output = self.mlp(layer_output)
        return layer_output
```
No need for all those white lines.

sgugger commented on src/transformers/models/perceiver/modeling_perceiver.py:
```diff
@@ -0,0 +1,2726 @@
+        # Construct a single block of self-attention layers.
+        # We get deeper architectures by applying this block more than once.
+        self_attends = []
+        for _ in range(config.num_self_attends_per_block):
+            layer = PerceiverLayer(
+                config,
+                is_cross_attention=False,
+                qk_channels=config.qk_channels,
+                v_channels=config.v_channels,
+                num_heads=config.num_self_attention_heads,
+                q_dim=config.d_latents,
+                kv_dim=config.d_latents,
+                widening_factor=config.self_attention_widening_factor,
+            )
+            self_attends.append(layer)
```
Can be done in one list comprehension?

sgugger commented on src/transformers/models/perceiver/modeling_perceiver.py:
```diff
@@ -0,0 +1,2726 @@
+
+        # Construct a single block of self-attention layers.
+        # We get deeper architectures by applying this block more than once.
+        self_attends = []
```
Should just be `layers`?

sgugger commented on src/transformers/models/perceiver/modeling_perceiver.py:
```diff
@@ -0,0 +1,2726 @@
+
+                layer_outputs = layer_module(
+                    hidden_states,
+                    attention_mask,
+                    layer_head_mask,
+                    output_attentions=output_attentions,
```
Like before, either use positional or keyword style but be consistent for all non-optional arguments.
Fits in one line if you pick positional style.

sgugger commented on src/transformers/models/perceiver/modeling_perceiver.py:
```diff
@@ -0,0 +1,2726 @@
+        if not return_dict:
+            return tuple(
+                v
+                for v in [
+                    hidden_states,
+                    all_hidden_states,
+                    all_self_attentions,
+                    all_cross_attentions,
+                ]
```
Fits in one line.

sgugger commented on src/transformers/models/perceiver/modeling_perceiver.py:
```diff
@@ -0,0 +1,2726 @@
+                logits = self.output_postprocessor(logits, modality_sizes=output_modality_sizes)
+
+        if not return_dict:
+            return (
+                logits,
+                sequence_output,
+            ) + encoder_outputs[1:]
```
Fits in one line.

sgugger commented on src/transformers/models/perceiver/modeling_perceiver.py:
```diff
@@ -0,0 +1,2726 @@
+                num_heads=8,
+                use_query_residual=False,
+                final_project=False,
+                trainable_position_encoding_kwargs=dict(
+                    num_channels=config.d_model, index_dims=config.max_position_embeddings
+                ),
```
This dictionary should be defined before you instantiate the decoder, for readability.

sgugger commented on src/transformers/models/perceiver/modeling_perceiver.py:
```diff
@@ -0,0 +1,2726 @@
+                position_encoding_type="trainable",
+                concat_or_add_pos="concat",
+                project_pos_dim=256,
+                trainable_position_encoding_kwargs=dict(
+                    num_channels=256,
+                    index_dims=config.image_size ** 2,
+                ),
```
Same for this dictionary

sgugger commented on src/transformers/models/perceiver/modeling_perceiver.py:
```diff
@@ -0,0 +1,2726 @@
+        logits = outputs.logits if return_dict else outputs[0]
+
+        loss = None
+        if labels is not None:
+            if self.num_labels == 1:
+                #  We are doing regression
+                loss_fct = MSELoss()
+                loss = loss_fct(logits.view(-1), labels.view(-1))
+            else:
+                loss_fct = CrossEntropyLoss()
+                loss = loss_fct(logits.view(-1, self.num_labels), labels.view(-1))
```
Use the problem type from the config to pick the best loss function (see what's happening in BertForSequenceClassification for instance).

sgugger commented on src/transformers/models/perceiver/modeling_perceiver.py:
```diff
@@ -0,0 +1,2726 @@
+                config,
+                prep_type="pixels",
+                spatial_downsample=1,
+                fourier_position_encoding_kwargs=dict(
+                    concat_pos=True, max_resolution=(224, 224), num_bands=64, sine_only=False
+                ),
```
This dict should be defined before for readability.

sgugger commented on src/transformers/models/perceiver/modeling_perceiver.py:
```diff
@@ -0,0 +1,2726 @@
+            decoder=PerceiverClassificationDecoder(
+                config,
+                num_channels=config.d_latents,
+                trainable_position_encoding_kwargs=dict(num_channels=config.d_latents, index_dims=1),
```
Same here.

sgugger commented on src/transformers/models/perceiver/modeling_perceiver.py:
```diff
@@ -0,0 +1,2726 @@
+        logits = outputs.logits if return_dict else outputs[0]
+
+        loss = None
+        if labels is not None:
+            if self.num_labels == 1:
+                #  We are doing regression
+                loss_fct = MSELoss()
+                loss = loss_fct(logits.view(-1), labels.view(-1))
+            else:
+                loss_fct = CrossEntropyLoss()
+                loss = loss_fct(logits.view(-1, self.num_labels), labels.view(-1))
```
This should use the problem type.

sgugger commented on src/transformers/models/perceiver/modeling_perceiver.py:
```diff
@@ -0,0 +1,2726 @@
+
+
+@add_start_docstrings(
+    """Example use of Perceiver for image classification, with Fourier position embeddings. """,
```
This could be expanded a bit on why it's different from `PerceiverForImageClassification`.

sgugger commented on src/transformers/models/perceiver/modeling_perceiver.py:
```diff
@@ -0,0 +1,2726 @@
+        )
+
+
+@add_start_docstrings("""Example use of Perceiver for image classification. """, PERCEIVER_START_DOCSTRING)
```
This docstring should be expanded.

sgugger commented on src/transformers/models/perceiver/modeling_perceiver.py:
```diff
@@ -0,0 +1,2726 @@
+                prep_type="conv",
+                spatial_downsample=1,
+                position_encoding_type="fourier",
+                fourier_position_encoding_kwargs=dict(
+                    concat_pos=True, max_resolution=(56, 56), num_bands=64, sine_only=False
+                ),
```
This dict should be defined before.

sgugger commented on src/transformers/models/perceiver/modeling_perceiver.py:
```diff
@@ -0,0 +1,2726 @@
+            decoder=PerceiverClassificationDecoder(
+                config,
+                num_channels=config.d_latents,
+                trainable_position_encoding_kwargs=dict(num_channels=config.d_latents, index_dims=1),
```
Same here.

sgugger commented on src/transformers/models/perceiver/modeling_perceiver.py:
```diff
@@ -0,0 +1,2726 @@
+        logits = outputs.logits if return_dict else outputs[0]
+
+        loss = None
+        if labels is not None:
+            if self.num_labels == 1:
+                #  We are doing regression
+                loss_fct = MSELoss()
+                loss = loss_fct(logits.view(-1), labels.view(-1))
+            else:
+                loss_fct = CrossEntropyLoss()
+                loss = loss_fct(logits.view(-1, self.num_labels), labels.view(-1))
+
```
Yes, like the two previous models :-)

sgugger commented on src/transformers/models/perceiver/modeling_perceiver.py:
```diff
@@ -0,0 +1,2726 @@
+                temporal_downsample=2,
+                position_encoding_type="fourier",
+                # position_encoding_kwargs
+                fourier_position_encoding_kwargs=dict(
+                    num_bands=64,
+                    max_resolution=config.train_size,
+                    sine_only=False,
+                    concat_pos=True,
+                ),
```
This dict should be defined before.

sgugger commented on src/transformers/models/perceiver/modeling_perceiver.py:
```diff
@@ -0,0 +1,2726 @@
+                # rather than a standard decoder position encoding.
+                position_encoding_type="fourier",
+                fourier_position_encoding_kwargs=dict(
+                    concat_pos=True, max_resolution=config.train_size, num_bands=64, sine_only=False
```
This one too.

sgugger commented on src/transformers/models/perceiver/modeling_perceiver.py:
```diff
@@ -0,0 +1,2726 @@
+        )
+
+
+@add_start_docstrings("""Example use of Perceiver for multimodal autoencoding. """, PERCEIVER_START_DOCSTRING)
```
This docstring should be expanded.

sgugger commented on src/transformers/models/perceiver/modeling_perceiver.py:
```diff
@@ -0,0 +1,2726 @@
+
+
+# Below: position encodings
+
+
```
```suggestion
```

sgugger commented on src/transformers/models/perceiver/modeling_perceiver.py:
```diff
@@ -0,0 +1,2726 @@
+
+
+# Below: Perceiver decoders
+
+
```
```suggestion
```

sgugger commented on src/transformers/models/perceiver/modeling_perceiver.py:
```diff
@@ -0,0 +1,2726 @@
+    def __init__(self, *args, **kwargs):
+        super(Conv2dSamePadding, self).__init__(*args, **kwargs)
+        self.zero_pad_2d = nn.ZeroPad2d(
+            reduce(__add__, [(k // 2 + (k - 2 * (k // 2)) - 1, k // 2) for k in self.kernel_size[::-1]])
```
This line does not make any sense, can we use a manual loop to compute the result?

sgugger commented on src/transformers/models/perceiver/modeling_perceiver.py:
```diff
@@ -0,0 +1,2726 @@
+    """
+    Generate a Fourier frequency position encoding with linear spacing.
+
+    Args:
+      pos: The Tensor containing the position of n points in d dimensional space.
+        A Torch tensor of shape [batch_size, n, d].
+      num_bands: The number of bands (K) to use.
+      max_resolution: The maximum resolution (i.e. the number of pixels per dim).
+        A tuple representing resolution for each dimension.
+      concat_pos: Whether to concatenate the input position encoding to the Fourier features.
+      sine_only: Whether to use a single phase (sin) or two (sin/cos) for each frequency band.
+
+    Returns:
+      embedding: A Torch tensor of shape [batch_size, n, n_channels]. If concat_pos is True and sine_only is False,
+      output dimensions are ordered as: [dim_1, dim_2, ..., dim_d, sin(pi*f_1*dim_1), ..., sin(pi*f_K*dim_1), ...,
+      sin(pi*f_1*dim_d), ..., sin(pi*f_K*dim_d), cos(pi*f_1*dim_1), ..., cos(pi*f_K*dim_1), ..., cos(pi*f_1*dim_d),
+      ..., cos(pi*f_K*dim_d)], where dim_i is pos[:, i] and f_k is the kth frequency band.
```
Adapt to our docstring style or remove.

sgugger commented on src/transformers/models/perceiver/modeling_perceiver.py:
```diff
@@ -0,0 +1,2726 @@
+
+def build_linear_positions(index_dims, output_range=(-1.0, 1.0)):
+    """
+    Generate an array of position indices for an N-D input array.
+
+    Args:
+      index_dims: The shape of the index dimensions of the input array.
+      output_range: The min and max values taken by each input index dimension.
+
+    Returns:
+      A Torch tensor of shape [index_dims[0], index_dims[1], .., index_dims[-1], N].
```
Same here.

sgugger commented on src/transformers/models/perceiver/modeling_perceiver.py:
```diff
@@ -0,0 +1,2726 @@
+    """
+    Checks or builds spatial position features (x, y, ...).
+
+    Args:
+      pos: None, or an array of position features. If None, position features
+        are built. Otherwise, their size is checked.
+      index_dims: An iterable giving the spatial/index size of the data to be
+        featurized.
+      batch_size: The batch size of the data to be featurized.
+
+    Returns:
+      An array of position features, of shape [batch_size, prod(index_dims)].
```
Same here.

sgugger commented on src/transformers/models/perceiver/modeling_perceiver.py:
```diff
@@ -0,0 +1,2726 @@
+        """
+        Constructor.
+
+        Args:
+          modalities: dict mapping modality name to post processor for that modality
+          input_is_dict: If True, input is assumed to be dictionary structured,
+            and outputs keep the same dictionary shape. If False, input is a tensor which is sliced up during
+            postprocessing by `modality_sizes`.
+          name: name of the module
```
Same here.

sgugger commented on src/transformers/models/perceiver/modeling_perceiver.py:
```diff
@@ -0,0 +1,2726 @@
+    ) -> Mapping[str, torch.Tensor]:
+        if not self.input_is_dict:
+            # Slice up modalities by their sizes.
+            assert modality_sizes is not None
```
No new assert in a modeling file, test and raise an explicipit error instead.

sgugger commented on src/transformers/models/perceiver/modeling_perceiver.py:
```diff
@@ -0,0 +1,2726 @@
+        """
+        Constructor.
+
+        Args:
+            modalities: dict mapping modality name to preprocessor
+            mask_probs: dict mapping modality name to masking probability of that
+                modality
+            min_padding_size: the minimum padding size for all modalities.
+                The final output will have num_channels equal to the maximum channels across all modalities plus
+                min_padding_size.
```
Adapt to our docstring style.

sgugger commented on src/transformers/models/perceiver/tokenization_perceiver.py:
```diff
@@ -0,0 +1,202 @@
+        following format:
+
+        - single sequence: ``X``
+        - pair of sequences: ``A </s> B </s>``
```
It doesn't look consistent with the output of this function.

sgugger commented on tests/test_modeling_perceiver.py:
```diff
@@ -0,0 +1,919 @@
+# limitations under the License.
+""" Testing suite for the PyTorch Perceiver model. """
+
+
```
```suggestion
```

sgugger commented on tests/test_modeling_perceiver.py:
```diff
@@ -0,0 +1,919 @@
+            if self.use_input_mask:
+                input_mask = random_attention_mask([self.batch_size, self.seq_length])
+        elif model_class.__name__ == "PerceiverForImageClassification":
+            config.d_model = 512
+            inputs = floats_tensor([self.batch_size, self.num_channels, self.image_size, self.image_size])
+        elif model_class.__name__ == "PerceiverForImageClassificationFourier":
+            config.d_model = 261
+            inputs = floats_tensor([self.batch_size, self.num_channels, self.image_size, self.image_size])
+        elif model_class.__name__ == "PerceiverForImageClassificationConvProcessing":
+            config.d_model = 322
+            inputs = floats_tensor([self.batch_size, self.num_channels, self.image_size, self.image_size])
+        elif model_class.__name__ == "PerceiverForOpticalFlow":
+            config.d_model = 322
+            inputs = floats_tensor([self.batch_size, 2, 27, self.train_size[0], self.train_size[1]])
+        elif model_class.__name__ == "PerceiverForMultimodalAutoencoding":
+            config.d_model = 409
```
Those are all values that are very high for tests.

sgugger commented on tests/test_modeling_perceiver.py:
```diff
@@ -0,0 +1,919 @@
+        )
+
+    def create_and_check_for_masked_lm(
+        self,
+        config,
+        inputs,
+        input_mask,
+        token_labels,
```
Fits in one line.

sgugger commented on tests/test_modeling_perceiver.py:
```diff
@@ -0,0 +1,919 @@
+        self.parent.assertEqual(result.logits.shape, (self.batch_size, self.seq_length, self.vocab_size))
+
+    def create_and_check_for_image_classification(
+        self,
+        config,
+        inputs,
+        input_mask,
+        image_labels,
```
Fits in one line.

sgugger commented on tests/test_modeling_perceiver.py:
```diff
@@ -0,0 +1,919 @@
+
+    def prepare_config_and_inputs_for_common(self):
+        config_and_inputs = self.prepare_config_and_inputs()
+        (
+            config,
+            inputs,
+            input_mask,
+            sequence_labels,
+            token_labels,
+        ) = config_and_inputs
```
Fits in one line.

sgugger commented on tests/test_modeling_perceiver.py:
```diff
@@ -0,0 +1,919 @@
+
+    def prepare_config_and_inputs_for_model_class(self, model_class):
+        config_and_inputs = self.prepare_config_and_inputs(model_class)
+        (
+            config,
+            inputs,
+            input_mask,
+            sequence_labels,
+            token_labels,
+        ) = config_and_inputs
```
Fits in one line.

sgugger commented on tests/test_modeling_perceiver.py:
```diff
@@ -0,0 +1,919 @@
+        self.model_tester = PerceiverModelTester(self)
+        self.config_tester = ConfigTester(self, config_class=PerceiverConfig, hidden_size=37)
+
+    # def _initialize_model(self, model_class, config):
+    #     if model_class.__name__ == "PerceiverForMultimodalAutoencoding":
+    #         model = model_class(config, subsampling=self.model_tester.subsampling)
+    #     else:
+    #         model = model_class(config)
+    #     return model
```
To clean up?

sgugger commented on tests/test_modeling_perceiver.py:
```diff
@@ -0,0 +1,919 @@
+            return
+
+        for model_class in self.all_model_classes:
+            if model_class in [
+                *get_values(MODEL_MAPPING),
+                PerceiverForOpticalFlow,
+                PerceiverForMultimodalAutoencoding,
+            ]:
```
Fits in one line.

sgugger commented on tests/test_modeling_perceiver.py:
```diff
@@ -0,0 +1,919 @@
+                # we expect to have 2 cross-attentions, namely one in the PerceiverEncoder, and one in PerceiverBasicDecoder
+                expected_num_cross_attentions = 2
+            else:
+                # todo
```
To finish before this is merged.

sgugger commented on tests/test_modeling_perceiver.py:
```diff
@@ -0,0 +1,919 @@
+            )
+
+        for model_class in self.all_model_classes:
+            if model_class.__name__ == "PerceiverModel":
+                continue
```
Same comment as Lysandre above.

sgugger commented on tests/test_modeling_perceiver.py:
```diff
@@ -0,0 +1,919 @@
+                            torch.allclose(
+                                set_nan_tensor_to_zero(tuple_object), set_nan_tensor_to_zero(dict_object), atol=1e-5
+                            ),
+                            msg=f"Tuple and dict output are not equal. Difference: {torch.max(torch.abs(tuple_object - dict_object))}. Tuple has `nan`: {torch.isnan(tuple_object).any()} and `inf`: {torch.isinf(tuple_object)}. Dict has `nan`: {torch.isnan(dict_object).any()} and `inf`: {torch.isinf(dict_object)}.",
```
Should be split in several lines to respect the 119 char limit.

sgugger commented on tests/test_modeling_perceiver.py:
```diff
@@ -0,0 +1,919 @@
+                recursive_check(tuple_output, dict_output)
+
+        for model_class in self.all_model_classes:
+            if model_class.__name__ == "PerceiverModel":
+                continue
```
Same comment as Lysandre above.

sgugger commented on tests/test_modeling_perceiver.py:
```diff
@@ -0,0 +1,919 @@
+        for model_class in self.all_model_classes:
+            if model_class.__name__ == "PerceiverModel":
+                continue
+            (
+                original_config,
+                inputs_dict,
+            ) = self.model_tester.prepare_config_and_inputs_for_model_class(model_class)
```
Fits in one line

sgugger commented on tests/test_modeling_perceiver.py:
```diff
@@ -0,0 +1,919 @@
+
+            if model_class.__name__ == "PerceiverForMultimodalAutoencoding":
+                for modality in outputs[0].keys():
+                    print("Processing modality:", modality)
```
```suggestion
```

sgugger commented on tests/test_modeling_perceiver.py:
```diff
@@ -0,0 +1,919 @@
+                    out_2 = outputs[0][modality].cpu().numpy()
+                    out_2[np.isnan(out_2)] = 0
+
+                    print("Shape of out 2:", out_2.shape)
+
```
```suggestion
```

sgugger reviewed (approved):
Thanks for adding this new model.

It would be nice if the doc page could explain the differences between the three different image classification models and when one should be used over the others. Also, it looks like the type of image classification is picked by the task attribute of the config, which does not make sense to me, as those are all performing the same tasks.

Then, the extra dep should be removed.

NielsRogge commented on docs/source/model_doc/perceiver.rst:
```diff
@@ -0,0 +1,228 @@
+
+Here's a TLDR explaining how Perceiver works:
+
+The main problem with the self-attention mechanism of the Transformer is that the time- and memory requirements scale
```
Sorry, that's how we write it in Dutch :)

NielsRogge committed:
Apply first batch of suggestions from code review

NielsRogge commented on src/transformers/models/perceiver/configuration_perceiver.py:
```diff
@@ -0,0 +1,162 @@
+        num_self_attends_per_block=26,
+        num_self_attention_heads=8,
+        num_cross_attention_heads=8,
+        qk_channels=None,
+        v_channels=None,
```
Some models (like the multimodal one) won't work if we set the qk_channels to the one of the language model.

Within the model, the qk_channels are automatically set equal to the dimension of the latents in case they are `None`.

LysandreJik replied:
OK!

NielsRogge committed:
Apply more suggestions from code review

NielsRogge committed:
Update docstrings

NielsRogge committed:
Define dicts beforehand for readability

NielsRogge committed:
Rename task to architecture in conversion script, include PerceiverModel in tests

NielsRogge committed:
Add print statements for debugging

NielsRogge committed:
Fix tests on GPU

NielsRogge committed:
Remove preprocessors, postprocessors and decoders from main init

NielsRogge committed:
Add integration test

NielsRogge committed:
Fix docs

patrickvonplaten commented on docs/source/model_doc/perceiver.rst:
```diff
@@ -0,0 +1,227 @@
+baseline on the GLUE language benchmark without the need for input tokenization and achieves state-of-the-art
+performance on Sintel optical flow estimation.*
+
+Here's a TLDR explaining how Perceiver works:
```
Super nice!

patrickvonplaten commented on docs/source/model_doc/perceiver.rst:
```diff
@@ -0,0 +1,227 @@
+Here's a TLDR explaining how Perceiver works:
+
+The main problem with the self-attention mechanism of the Transformer is that the time and memory requirements scale
+quadratically in the sequence length. Hence, models like BERT and RoBERTa are limited to a max sequence length of 512
```
(nit)
```suggestion
quadratically with the sequence length. Hence, models like BERT and RoBERTa are limited to a max sequence length of 512
```

patrickvonplaten commented on docs/source/model_doc/perceiver.rst:
```diff
@@ -0,0 +1,227 @@
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. autoclass:: transformers.PerceiverForMultimodalAutoencoding
+    :members: forward
```
Super nice doc!

patrickvonplaten commented on src/transformers/models/perceiver/configuration_perceiver.py:
```diff
@@ -0,0 +1,161 @@
+        max_position_embeddings (:obj:`int`, `optional`, defaults to 2048):
+            The maximum sequence length that this model might ever be used with. Typically set this to something large
+            just in case (e.g., 512 or 1024 or 2048).
+        train_size (:obj:`List[int]`, `optional`, defaults to [368, 496]):
```
If it's just used for the optimal flow model - might it make sense to let this default to None to not confuse readers of the config of models that have nothing to do with optical flow?

patrickvonplaten commented on src/transformers/models/perceiver/modeling_perceiver.py:
```diff
@@ -0,0 +1,2769 @@
+
+
+@dataclass
+class PerceiverMaskedLMOutput(MaskedLMOutput):
```
not a fan of this abstraction - why not copy paste it here? I haven't seen this being done before - what do you think @sgugger ?

sgugger replied:
It only takes the copy of the four arguments, so it's easy to add here without depending on `MaskedLMOutput`, but I think we discussed with @LysandreJik that it would be better to have the model outputs use inheritances so we can make some checks on the classes. I don't remember the specific however.

NielsRogge replied:
@patrickvonplaten I've actually done this for quite some models now, e.g. [TAPAS](https://github.com/huggingface/transformers/blob/e4c67d60ec07ac6402951848c0e65639e81b65fe/src/transformers/models/tapas/modeling_tapas.py#L106), [CANINE](https://github.com/huggingface/transformers/blob/e4c67d60ec07ac6402951848c0e65639e81b65fe/src/transformers/models/canine/modeling_canine.py#L66)

NielsRogge replied:
Oh I see, you mean it's better to copy all attributes, I can do that.

patrickvonplaten replied:
I would prefer this, but let's see what @LysandreJik says

NielsRogge replied:
I've fixed it, copied all arguments.

patrickvonplaten commented on src/transformers/models/perceiver/modeling_perceiver.py:
```diff
@@ -0,0 +1,2769 @@
+
+
+@dataclass
+class PerceiverClassifierOutput(SequenceClassifierOutput):
```
same here

patrickvonplaten commented on src/transformers/models/perceiver/modeling_perceiver.py:
```diff
@@ -0,0 +1,2769 @@
+
+    def __init__(self, config):
+        super().__init__()
+        self.latents = nn.Parameter(torch.randn(config.num_latents, config.d_latents))
```
Think in general we try to do the initialization part on the `_init_weights()` function. But no big deal for me

patrickvonplaten commented on src/transformers/models/perceiver/modeling_perceiver.py:
```diff
@@ -0,0 +1,2769 @@
+        self.latents = nn.Parameter(torch.randn(config.num_latents, config.d_latents))
+
+    def forward(self, batch_size):
+        return self.latents.expand(batch_size, -1, -1)  # Thanks, Phil Wang
```
haha nice comment

patrickvonplaten replied:
I wonder whether he will ever see this

NielsRogge replied:
cc @lucidrains

patrickvonplaten replied:
Well that's a bit too easy now

patrickvonplaten commented on src/transformers/models/perceiver/modeling_perceiver.py:
```diff
@@ -0,0 +1,2769 @@
+    config_class = PerceiverConfig
+    base_model_prefix = "perceiver"
+
+    def _init_weights(self, module):
```
I would also init the latents of the `PerceiverEmbeddings` here instead of directly in the module

patrickvonplaten commented on src/transformers/models/perceiver/modeling_perceiver.py:
```diff
@@ -0,0 +1,2769 @@
+    PERCEIVER_START_DOCSTRING,
+)
+class PerceiverModel(PerceiverPreTrainedModel):
+    def __init__(self, config, decoder=None, input_preprocessor=None, output_postprocessor=None):
```
Cool! That's a very nice solution to the complexity of the input modality! 

Could we however mabye add a docstring and type hint for both `input_preprocessor` and `output_postprocessor` here? This would be extremely useful for the reader IMO

patrickvonplaten commented on src/transformers/models/perceiver/tokenization_perceiver.py:
```diff
@@ -0,0 +1,202 @@
+logger = logging.get_logger(__name__)
+
+
+class PerceiverTokenizer(PreTrainedTokenizer):
```
Gently passing this on to @Narsil who understands byte-level tokenizer way better than me

patrickvonplaten reviewed (approved):
Amazing PR! Really impressed how this super complex model was made so easy to read! 

I strongly agree with @LysandreJik and @sgugger regarding not adding a new `einops` dependency here. 

Besides that I'm not super happy with abstracting model output classes as I don't think this is common practice in `transformers` and I think it goes a bit against (everything in one file philosophy) - what do you think @sgugger here? 

The rest is only nits

NielsRogge committed:
Replace einops by torch

NielsRogge committed:
Update for new docs frontend

NielsRogge commented:
Thanks for the reviews, I've:

* removed the einops dependency, and replaced it by just PyTorch.
* fixed the tests on GPU
* renamed `PerceiverForImageClassification` to `PerceiverForImageClassificationLearned` (to better reflect that this one uses learned position embeddings, which is more in line with the other models, namely `PerceiverForImageClassification` and `PerceiverForImageClassificationConvProcessing`)

To do: 
- now, it's mainly about writing more docstrings.

NielsRogge committed:
Rename PerceiverForImageClassification

NielsRogge committed:
Improve docs

NielsRogge committed:
Improve docs

NielsRogge committed:
Improve docs of PerceiverModel

NielsRogge committed:
Fix some more tests

NielsRogge committed:
Improve center_crop

NielsRogge committed:
Add PerceiverForSequenceClassification

tonibagur commented:
Hi, reading this nice perceiver io implementation I came up with a question: It seems that in Perceiver the input is iteratively mixed with the latents via cross attention but in Perceiver IO it seems that the input is fed only once in the network. Does somebody know if I am right and the reason of doing it in this way? If I got it right, it seems that also in this implementation the input is fed only once...

esceptico commented:
Hi @tonibagur 
As mentioned in the paper:

>  The encode and processing stages re-use the basic structure of the original Perceiver. With the exception of the repeated encoder cross-attends. We found it simpler to omit these modules, as
 they increase compute without dramatically increasing performance, as reported in Appendix Table 6 in original Perceiver paper.

tonibagur commented:
> Hi @tonibagur As mentioned in the paper:
> 
> > The encode and processing stages re-use the basic structure of the original Perceiver. With the exception of the repeated encoder cross-attends. We found it simpler to omit these modules, as
> > they increase compute without dramatically increasing performance, as reported in Appendix Table 6 in original Perceiver paper.

Thanks esceptico, I completely missed this paragraph :'

NielsRogge committed:
Small improvements

NielsRogge committed:
Fix tests

NielsRogge committed:
Add integration test for optical flow model

Narsil commented on src/transformers/models/perceiver/tokenization_perceiver.py:
```diff
@@ -0,0 +1,202 @@
+    def _convert_id_to_token(self, index):
+        """Converts an index (integer) in a token (str) using the vocab."""
+        if index in self.special_tokens_decoder:
+            token = self.special_tokens_decoder[index]
```
Isn't this missing the `added_tokens` support ?

NielsRogge replied:
Can you clarify? There's a `special_tokens_decoder` which includes the special tokens.

Narsil replied:
`added_tokens` are added by users.

`tokenizer.add_tokens(["[MYSPECIAL_TOKEN]"])` and enables adding stuff to an existing tokenizer.

This would add token (256 + 5) (and someone could add any number of those).

They seem to be  handled properly in `convert_tokens_to_id` so they should probably be handled here too.

What's odd, is that the automated tests should test for this.

NielsRogge replied:
Ok, I've added it.

Narsil commented on src/transformers/models/perceiver/tokenization_perceiver.py:
```diff
@@ -0,0 +1,202 @@
+            else:
+                tok_string = bytes([ord(token)])
+            bstring += tok_string
+        string = bstring.decode("utf-8")
```
This was changed in `ByT5` to use `errors="ignore"` to prevent impossible bytes to be wrongly decoded.

I personally prefer crashing than ignoring stuff, but mentioning in case you need the same behavior.
Ultimately, it's true that crashing is a bit strong. I don't have a lot of overview of perceiver, but if it's supposed to output bytes for something else than utf-8 text, maybe something else is required.

NielsRogge replied:
I've added `errors="replace"` to reflect the [original implementation](https://github.com/deepmind/deepmind-research/blob/master/perceiver/bytes_tokenizer.py).

Narsil reviewed (commented):
This overall looks ok, but there are no tests for the tokenizer ?

Some comments.

NielsRogge committed:
Clean up

NielsRogge committed:
Add tests for tokenizer

Narsil commented on tests/test_tokenization_perceiver.py:
```diff
@@ -0,0 +1,284 @@
+        self.assertEqual(decoded, "e è é ê ë")
+
+        # encode/decode, but with `encode` instead of `__call__`
+        self.assertEqual(tokenizer.decode(tokenizer.encode("e è é ê ë")), "e è é ê ë")
```
Maybe another small test to test invalid byte sequence decoding.

tokenizer.decode([178]) should be enough I think. (178 > 128, so it's has the continuation bytes, and yet we don't continue).

It should show something like `�` when using `errors="replace"`.
Making it a test should help keep forward compatibility (maybe with a link to https://github.com/deepmind/deepmind-research/blob/master/perceiver/bytes_tokenizer.py#L30 )

Narsil reviewed (commented):
Last nit on my end.

NielsRogge committed:
Fix tokenizer by adding special tokens properly

Narsil reviewed (commented):
LGTM

NielsRogge committed:
Fix CI

LysandreJik reviewed (approved):
Great work, @NielsRogge !
